src/jit/morph.cpp

```diff
@@ -14455,7 +14455,10 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
 
                 /* Make sure these are all ints and precision is not lost */
 
-                if (cast >= dstt && dstt <= TYP_INT && srct <= TYP_INT)
+                unsigned sizeDstt = genTypeSize(dstt);
+                unsigned sizeInt  = genTypeSize(TYP_INT);
+
+                if (genTypeSize(cast) >= sizeDstt && sizeDstt <= sizeInt && genTypeSize(srct) <= sizeInt)
```

Comment:
The first condition is correct but the other 2 are not correct, they need to stay as there were. They're checking if we're dealing with integer types. The new conditions will accept `TYP_FLOAT` as well.